src/sentry/replays/usecases/query/conditions/aggregate.py

```diff
@@ -61,65 +61,85 @@ def visit_not_in(expression: Expression, value: list[int]) -> Condition:
     @staticmethod
     def visit_eq(expression: Expression, value: str | None) -> Condition:
+        if value is None:
+            # "All rows in aggregation set are null."
+            return does_not_contain(IPv4Scalar.visit_neq(expression, None))
         return contains(IPv4Scalar.visit_eq(expression, value))
 
     @staticmethod
     def visit_neq(expression: Expression, value: str | None) -> Condition:
+        if value is None:
+            # "1+ rows in aggregation set are non-null."
+            return contains(IPv4Scalar.visit_neq(expression, None))
         return does_not_contain(IPv4Scalar.visit_eq(expression, value))
 
     @staticmethod
     def visit_in(expression: Expression, value_list: list[str | None]) -> Condition:
+        nonnull_contains = contains(
+            IPv4Scalar.visit_in(expression, [v for v in value_list if v is not None])
+        )
         if None in value_list:
-            contains_cond = contains(
-                IPv4Scalar.visit_in(expression, [v for v in value_list if v is not None])
-            )
-            return Or(
-                conditions=[
-                    contains_cond,
-                    Condition(Function("isNull", parameters=[expression]), Op.EQ, 1),
-                ]
-            )
-        return contains(IPv4Scalar.visit_in(expression, value_list))
+            return Or(conditions=[SumOfIPv4Scalar.visit_eq(expression, None), nonnull_contains])
+        return nonnull_contains
 
     @staticmethod
     def visit_not_in(expression: Expression, value_list: list[str | None]) -> Condition:
+        nonnull_does_not_contain = does_not_contain(
+            IPv4Scalar.visit_in(expression, [v for v in value_list if v is not None])
+        )
         if None in value_list:
-            does_not_contain_cond = does_not_contain(
-                IPv4Scalar.visit_in(expression, [v for v in value_list if v is not None])
-            )
             return And(
-                conditions=[
-                    does_not_contain_cond,
-                    Condition(Function("isNull", parameters=[expression]), Op.EQ, 0),
-                ]
+                conditions=[SumOfIPv4Scalar.visit_neq(expression, None), nonnull_does_not_contain]
             )
-        return does_not_contain(IPv4Scalar.visit_in(expression, value_list))
+        return nonnull_does_not_contain
 
 
 class SumOfStringScalar(GenericBase):
     @staticmethod
     def visit_eq(expression: Expression, value: str) -> Condition:
+        if value == "":
+            # "All rows in aggregation set are null."
+            return does_not_contain(StringScalar.visit_neq(expression, ""))
         return contains(StringScalar.visit_eq(expression, value))
 
     @staticmethod
     def visit_neq(expression: Expression, value: str) -> Condition:
+        if value == "":
+            # "1+ rows in aggregation set are non-null."
+            return contains(StringScalar.visit_neq(expression, ""))
```

Comment:
This is hard to read.  Abstract `StringScalar.visit_neq(expression, "")` to a function called `filled(expr)` so you can write `contains(filled(expr))` and `does_not_contain(filled(expr))`.  Apply everywhere we're doing this empty check.

Reply:
The name `filled` is unclear to me. How about `nonempty` or `is_nonempty`?

Reply:
Sure.